homeassistant/components/icloud.py

```diff
@@ -0,0 +1,956 @@
+            self._distance = distance(device_state_lat, device_state_long,
+                                      zone_state_lat, zone_state_long)
+            self._distance = round(self._distance / 1000, 1)
+        if 'battery' in new_state.attributes:
+            self._battery = new_state.attributes['battery']
+
+        if new_state.state == self._overridestate:
+            self.update_ha_state()
+            return
+
+        self._overridestate = None
+
+        if new_state.state != 'not_home':
+            self._interval = 30
+            self.update_ha_state()
+        else:
+            if self._distance is None:
+                self.update_ha_state()
+                return
+            if self._distance > 100:
+                self._interval = round(self._distance, 0)
+                if self._googletraveltime is not None:
+                    gttstate = self.hass.states.get(self._googletraveltime)
+                    if gttstate is not None:
+                        self._interval = round(float(gttstate.state) - 10, 0)
+            elif self._distance > 50:
+                self._interval = 30
+            elif self._distance > 25:
+                self._interval = 15
+            elif self._distance > 10:
+                self._interval = 5
+            else:
+                self._interval = 1
+            if self._battery is not None:
+                if self._battery <= 33 and self._distance > 3:
+                    self._interval = self._interval * 2
+            self.update_ha_state()
+
+
+class IEvent(Entity):  # pylint: disable=too-many-instance-attributes
+    """Represent an icloud calendar event in Home Assistant."""
+
+    def __init__(self, hass, icloudobject, name, typeevent=None):
+        """Initialize an iEvent."""
+        # pylint: disable=too-many-arguments
+        self.hass = hass
+        self.icloudobject = icloudobject
+        self.api = icloudobject.api
+        self.eventguid = name
+        self._starttime = None
+        self._starttext = None
+        self._endtime = None
+        self._endtext = None
+        self._duration = None
+        self._title = None
+        self._remaining = 0
+        self._remainingtext = None
+        self._location = None
+        self._type = typeevent
+        self._tz = None
+
+        self.entity_id = generate_entity_id(
+            ENTITY_ID_FORMAT_EVENT, self.eventguid,
+            hass=self.hass)
+
+    @property
+    def state(self):
+        """Return the state of the icloud event."""
+        return self._title
+
+    @property
+    def state_attributes(self):
+        """Return the attributes of the icloud event."""
+        return {
+            ATTR_STARTTIME: self._starttext,
+            ATTR_ENDTIME: self._endtext,
+            ATTR_DURATION: self._duration,
+            ATTR_REMAINING: self._remainingtext,
+            ATTR_LOCATION: self._location,
+            ATTR_FRIENDLY_NAME: self._type
```

Comment:
Here we can use the same `dict` approach I suggested above.

The main reason for this is that all these attributes are only displayed through the `state_attributes` function. We might as well put it in the correct format while we read the data